File: lightning.cpp
#include <algorithm>
#include <assert.h>
#include <stdio.h>
#include <string>
#include <unistd.h>

struct Buffer {
  char data[BUFSIZ];
  size_t index = 0;

  void extend(const char *str, size_t size) {
    std::copy(str, str + size, data + index);
    index += size;
  }

  void push(char c) { data[index++] = c; }

  void clear() { index = 0; }

  size_t size() const { return index; }
};

int increase_str_num(char *input, int input_size) {
  int i = input_size - 1;
  for (; i >= 0; --i) {
    ++input[i];
    if (input[i] - '0' < 10)
      return input_size - i;
    input[i] = '0';
  }
  assert(!"overflow input");
  return 0;
}

int main() {
  Buffer buffer;
  char num[12] = "00000000000";
  int num_len = 1;
  const char prefix[] = "Hello, ";
  const size_t LINE_MAX_LEN = 17;
  for (unsigned int i = 0; i < 100'000'000; ++i) {
    buffer.extend(prefix, sizeof(prefix) - 1);
    buffer.extend(num + sizeof(num) - 1 - num_len, num_len);
    num_len = std::max(num_len, increase_str_num(num, sizeof(num) - 1));
    buffer.push('\t');
    if (buffer.size() + LINE_MAX_LEN > BUFSIZ) {
      write(STDOUT_FILENO, buffer.data, buffer.size());
      buffer.clear();
    }
  }
  write(STDOUT_FILENO, buffer.data, buffer.size());
}

File: lightning.rs
pub struct Buffer {
    pub bytes: Vec<u8>,
    pub fd: std::os::unix::io::RawFd,
}

impl Buffer {
    #[inline(always)]
    pub fn new(
        byte_capacity: usize,
        fd: std::os::unix::io::RawFd,
    ) -> Self {
        Self {
            bytes: Vec::with_capacity(byte_capacity),
            fd,
        }
    }

    #[inline(always)]
    pub fn flush(&mut self) {
        let _ = nix::unistd::write(self.fd, &self.bytes).unwrap();
    }

    #[inline(always)]
    pub fn clear(&mut self) {
        self.bytes.clear();
    }

    #[inline(always)]
    pub fn enqueue(&mut self, bytes: &[u8]) {
        self.bytes.extend_from_slice(bytes);
    }
}

#[inline(always)]
fn push_hello(buf: &mut Buffer, num: u32) {
    buf.enqueue(b"\tHello, ");
    let _ = itoa::write(&mut buf.bytes, num).unwrap();
}

fn main() {
    const BUFSIZ: usize = libc::BUFSIZ as usize;
    const TARGET: u32 = 100_000_000;
    let mut buf = Buffer::new(BUFSIZ, libc::STDOUT_FILENO);
    buf.enqueue(b"Hello, 0");
    let mut i = 1;
    const PREFIX_SIZE: usize = 1 + 7; // b"\tHello, ".len();
    const NUM_LEN: usize = 9; // b"100000000".len();
    'done: while i < TARGET {
        while i < TARGET
            && BUFSIZ > buf.bytes.len() + PREFIX_SIZE + NUM_LEN
        {
            push_hello(&mut buf, i);
            i += 1;
            if i >= TARGET {
                break 'done;
            }
        }
        buf.flush();
        buf.clear();
        push_hello(&mut buf, i);
        i += 1;
    }
    buf.enqueue(b"\t");
    buf.flush();
}
